google-cloud-spanner/lib/google/cloud/spanner/results.rb

```diff
@@ -98,24 +101,71 @@ def rows
               @metadata ||= grpc.metadata
               @stats ||= grpc.stats
 
-              if chunked_value
-                grpc.values.unshift merge(chunked_value, grpc.values.shift)
-                chunked_value = nil
-              end
-              to_iterate = values + grpc.values
-              chunked_value = to_iterate.pop if grpc.chunked_value
+              buffered_responses << grpc
+
+              if (grpc.resume_token && grpc.resume_token != "") ||
+                buffered_responses.size >= buffer_upper_bound
+                # This can set the resume_token to nil
```

Comment:
This is (understandably) involved, and either comments or refactoring to smaller methods with meaningful names will help make it easier to follow. Just adding some comments will be fine for now.